ql/src/java/org/apache/hadoop/hive/ql/queryhistory/repository/IcebergRepository.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.hadoop.hive.ql.QueryState;
+import org.apache.hadoop.hive.ql.exec.FileSinkOperator.RecordWriter;
+import org.apache.hadoop.hive.ql.exec.Utilities;
+import org.apache.hadoop.hive.ql.io.HiveFileFormatUtils;
+import org.apache.hadoop.hive.ql.io.HiveOutputFormat;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.metadata.HiveStorageHandler;
+import org.apache.hadoop.hive.ql.metadata.HiveUtils;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.hadoop.hive.ql.plan.TableDesc;
+import org.apache.hadoop.hive.ql.queryhistory.schema.QueryHistoryRecord;
+import org.apache.hadoop.hive.ql.queryhistory.schema.QueryHistorySchema;
+import org.apache.hadoop.hive.ql.security.authorization.HiveCustomStorageHandlerUtils;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.hive.ql.session.SessionStateUtil;
+import org.apache.hadoop.hive.serde2.Serializer;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.TaskAttemptID;
+import org.apache.tez.mapreduce.hadoop.mapred.TaskAttemptContextImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IcebergRepository extends AbstractQueryHistoryRepository implements QueryHistoryRepository {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergRepository.class);
+  private static final String ICEBERG_STORAGE_HANDLER = "org.apache.iceberg.mr.hive.HiveIcebergStorageHandler";
+
+  private HiveOutputFormat<?, ?> outputFormat;
+  private Serializer serializer;
+  @VisibleForTesting
+  HiveStorageHandler storageHandler;
+  @VisibleForTesting
+  TableDesc tableDesc;
+
+  @Override
+  public void init(HiveConf conf, QueryHistorySchema schema) {
+    super.init(conf, schema);
+  }
+
+  @Override
+  protected Table createTable(Hive hive, Database db) throws HiveException {
+    LOG.info("Creating iceberg Query History table...");
+
+    Table table = getInitialTable();
+
+    // iceberg specific
+    table.setProperty(org.apache.hadoop.hive.metastore.api.hive_metastoreConstants.META_TABLE_STORAGE,
+        ICEBERG_STORAGE_HANDLER);
+    table.setProperty("table_type", "ICEBERG");
+    table.setProperty("write.format.default", "orc");
+    table.setProperty(hive_metastoreConstants.META_TABLE_NAME, QUERY_HISTORY_DB_TABLE_NAME);
+
+    table.setFields(schema.getFields());
+
+    hive.createTable(table, false);
+
+    table = hive.getTable(QUERY_HISTORY_DB_NAME, QUERY_HISTORY_TABLE_NAME);
+    table.setProperty("iceberg.mr.table.location", table.getDataLocation().toString());
+
+    return table;
+  }
+
+  @Override
+  protected void postInitTable(Table table) throws Exception {
+    this.tableDesc = Utilities.getTableDesc(table);
+
+    Map<String, String> map = new HashMap<>();
+    this.storageHandler = HiveUtils.getStorageHandler(conf, tableDesc.getProperties()
```

Comment:
can't we use `table.getStorageHandler()`?

Reply:
oh, I didn't know this one, thanks